### DISCUSSION TITLE
Using OpenAI ChatGPT to generate HTML documentation for Pascal source code

### DISCUSSION THREAD

**Discussion by michaliskambi** [pasdoc/pasdoc#170] [General]
Yeah. AI can do the job of PasDoc, basically. 

See my post: [My mind is blown. I can use AI to generate Castle Game Engine code to integrate it with PhysX. I can use AI to generate HTML documentation from comments in Pascal code.](https://castle-engine.io/wp/2022/12/17/my-mind-is-blown-i-can-use-ai-to-generate-castle-game-engine-code-to-integrate-it-with-physx-i-can-use-ai-to-generate-html-documentation-from-comments-in-pascal-code/) .

Wow.

**Comment by Fr0sT-Brutal** [comment]
Looks really impressive but does the code make any sense really? Answers of that bot were recently banned at SO just for the reason they might look quite useful but are not

**Reply by michaliskambi** [reply]
Well it's far from flawless. 

It can indeed say invalid things (and it will say them with the same confidence as when it says valid things :), and to judge it you have to just query the Internet -- it will not provide you direct URLs to confirm this-or-that claim). E.g. it said that Castle Game Engine supports PlayStation (which we never did) likely because Internet search for "Castle Game Engine PlayStation" indeed finds something (a project on PlayStation called "the Castle Game" and unrelated to "Castle Game Engine").

Still, from my experiments -- it is impressive, and the answers are a useful starting point. Like doing Internet search (DuckDuckGo / Google etc.), but sometimes much better, because it can "synthesize" knowledge that isn't directly spelled out in any particular webpage.

I was testing it to solve some issues with Jenkins, I was testing it to discover the API of PhysX, I was testing it to write a bot for Discord and to write Python code that interacts with OpenAI. Nothing was flawless... but it gave me a significant insight into what to follow next.

And in case like "Integrating Castle Game Engine with PhysX" -- wow. I shown it on https://castle-engine.io/wp/2022/12/17/my-mind-is-blown-i-can-use-ai-to-generate-castle-game-engine-code-to-integrate-it-with-physx-i-can-use-ai-to-generate-html-documentation-from-comments-in-pascal-code/ . Although the code there isn't complete, it doesn't "just compile"... but it's a great start, a start much better than searching the Internet for "Castle Game Engine PhysX" or "Pascal PhysX" gives you (because nobody has yet actually done it in Pascal, there are no solutions for it). Sure I have to read PhysX SDK docs / examples anyway, to confirm that the approach makes sense. Still, it's great to have such "starting point" :)

And in case of "simulating PasDoc" (generate HTML from Pascal)... it's sensible. Although it's ugly HTML (though I could probably ask it to add some CSS to it) but it makes sense. And it did it without knowing anything from me about Pascal, about my coding conventions, I never told it "before each Pascal identifier, there's a comment documenting it". And it figured out to split class description into sections, like "Constants", "Procedures", "Properties".

I do not yet know how much it will really change the way I work. But it is surely fun to test it on real questions :) and get insights that a simple Internet search would not give.
